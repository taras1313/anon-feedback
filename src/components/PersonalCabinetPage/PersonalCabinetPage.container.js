import { connect } from 'react-redux';

import component from './PersonalCabinetPage';

const mapStateToProps = state => {
	const {
		userReducer: { user }
  } = state;
  
  return {
    user
  };
};

export const PersonalCabinetPage = connect(mapStateToProps)(component);
