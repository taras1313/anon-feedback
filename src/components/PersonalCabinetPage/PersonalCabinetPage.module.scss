.buttonContainer {
	width: 100%;
	display: flex;
	justify-content: center;
	height: 60px;
	padding: 10px 0px;
}

.active {
  color: white !important;
  background-color: #3f51b5 !important;
}
