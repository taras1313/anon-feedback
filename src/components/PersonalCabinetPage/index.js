export * from './PersonalCabinetPage.container.js'
