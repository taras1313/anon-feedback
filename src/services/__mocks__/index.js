export const threadService = {
  getSubscribed: jest.fn(() => Promise.resolve([{}])),
  getLiked: jest.fn(() => Promise.resolve([{}])),
  getCreated: jest.fn(() => Promise.resolve([{}])),
  getReplied: jest.fn(() => Promise.resolve([{}])),
  createThread: jest.fn((thread) => Promise.resolve(thread)),
  updateThread: jest.fn((thread) => Promise.resolve(thread)),
  createComment: jest.fn((comment) => Promise.resolve(comment)),
  updateComment: jest.fn((comment) => Promise.resolve(comment)),
  getThread: jest.fn((data) => Promise.resolve(data)), 
  subscribeToThread: jest.fn((data) => Promise.resolve(data)), 
  unsubscribeFromThread: jest.fn((data) => Promise.resolve(data)), 
  getThreads: jest.fn(() => Promise.resolve([{}])),
  like: jest.fn((data) => Promise.resolve(data)), 
  dislike: jest.fn((data) => Promise.resolve(data)), 
  likeComment: jest.fn((data) => Promise.resolve(data)), 
  dislikeComment: jest.fn((data) => Promise.resolve(data)), 
};
