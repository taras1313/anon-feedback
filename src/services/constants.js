export const baseUrlDev = 'http://localhost:5000';
export const baseUrlProd = 'add here url for prod';
export const pathList = {
  user: {
    base: '/users',
    login: '/users/login'
  },
  thread: {
    base: '/threads',
    like: '/threads/like',
    dislike: '/threads/dislike'
  }
};
