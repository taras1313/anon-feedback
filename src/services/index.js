import { ApiAdapter } from './api.adapter';
import { UserService } from './UserService';
import { ThreadService } from './ThreadService';

const apiAdapter = new ApiAdapter();
export const userService = new UserService(apiAdapter);
export const threadService = new ThreadService(apiAdapter);
