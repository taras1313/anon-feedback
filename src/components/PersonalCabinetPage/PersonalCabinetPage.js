import React, { Component } from 'react';

import ButtonGroup from '@material-ui/core/ButtonGroup';
import Button from '@material-ui/core/Button';
import { threadService } from '../../services';
import Loader from '../Loader/Loader';
import { ThreadComponent } from '../TreadComponent';

import styles from './PersonalCabinetPage.module.scss';

const SUBSCRIBED = 'SUBSCRIBED';
const LIKED = 'LIKED';
const CREATED = 'CREATED';
const REPLIED = 'REPLIED';

const threadOptions = [SUBSCRIBED, LIKED, CREATED, REPLIED];

export default class PersonalCabinetPage extends Component {
	state = {
		threads: [],
		option: SUBSCRIBED
	};

	componentDidMount() {
		const {
			user: { subscribedThreads }
		} = this.props;
		threadService.getSubscribed(subscribedThreads).then(res => {
			this.setState({ threads: res });
		});
	}

	renderThreads() {
		const { threads } = this.state;
		return threads.map(el => <ThreadComponent key={el._id} threadView="preview" thread={el} />);
	}

	optionChangeHandler = option => {
		const {
			user: { createdThreads, subscribedThreads, _id, repliedList }
		} = this.props;

		console.log('subbed threads', subscribedThreads);
		const switchOption = () => {
			switch (option) {
				case SUBSCRIBED:
					return threadService.getSubscribed(subscribedThreads);

				case LIKED:
					return threadService.getLiked(_id);

				case CREATED:
					return threadService.getCreated(createdThreads);

				case REPLIED:
					const repliedArr = repliedList.map(({ threadId }) => threadId);
					return threadService.getReplied(repliedArr);

				default:
					break;
			}
		};

		switchOption().then(res => {
			console.log('res', res);
			this.setState({ threads: res, option });
		});
	};

	isActive = label => this.state.option === label;

	renderButtons() {
		return threadOptions.map(label => (
			<Button
				key={label}
				onClick={() => this.optionChangeHandler(label)}
				name={label}
				className={this.isActive(label) ? styles.active : ''}
			>
				{label}
			</Button>
		));
	}

	render() {
		const { threads } = this.state;

		return (
			<>
				<div className={styles.buttonContainer}>
					<ButtonGroup size="small" aria-label="small outlined button group" color="primary">
						{this.renderButtons()}
					</ButtonGroup>
				</div>
				{threads.length ? this.renderThreads() : <Loader />}
			</>
		);
	}
}
