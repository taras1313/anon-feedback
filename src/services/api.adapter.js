/* eslint-disable */
import axios from 'axios';
import { baseUrlDev, baseUrlProd } from './constants';

export class ApiAdapter {
  constructor() {
    const env = process.env.REACT_APP_ENV;
    this._url = env === 'development' ? baseUrlDev : baseUrlProd;

    // alternative to approach above
    // this._url = location.origin.includes('localhost') ? baseUrlDev : baseUrlProd;
  }

  composeUrl = path => `${this._url}${path}`;

  get = (path, params = {}) => {
    const endpoint = this.composeUrl(path);

    const request = axios.get(endpoint, { params })
      .then(response => response)
      .catch(error => Promise.reject({ error }));

    return request;
  };

  post = (path, body = {}) => {
    const endpoint = this.composeUrl(path);
    
    const request = axios.post(endpoint, body)
      .then(response => {
        return response;
      })
      .catch(error => Promise.reject({ error }));

    return request;
  };

  put = (path, body = {}) => {
    const endpoint = this.composeUrl(path);

    const request = axios.put(endpoint, body)
      .then(response => response)
      .catch(error => Promise.reject({ error }));

    return request;
  };

  patch = (path, body = {}) => {
    const endpoint = this.composeUrl(path);

    const request = axios.patch(endpoint, body)
      .then(response => response)
      .catch(error => Promise.reject({ error }));

    return request;
  };
}
