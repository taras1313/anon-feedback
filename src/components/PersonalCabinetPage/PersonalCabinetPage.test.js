import React from "react";
import { shallow } from "enzyme";
import Button from "@material-ui/core/Button";
import PersonalCabinetPage from "./PersonalCabinetPage";
jest.mock("../../services");

// eslint-disable-next-line import/first
import { threadService } from "../../services";

describe("Personal cabinet page", () => {
  const props = { user: { repliedList: [{ threadId: 1 }] } };

  afterEach(() => {
    jest.clearAllMocks();
  });

  test("snapshot", () => {
    const wrapper = shallow(<PersonalCabinetPage {...props} />);
    expect(wrapper).toMatchSnapshot();
    expect(threadService.getSubscribed).toHaveBeenCalled();
  });

  test("gets subscribed threads", () => {
    const wrapper = shallow(<PersonalCabinetPage {...props} />);
    wrapper.find(Button).at(0).simulate("click");
    expect(threadService.getSubscribed).toHaveBeenCalledTimes(2);
  });

  test("gets liked threads", () => {
    const wrapper = shallow(<PersonalCabinetPage {...props} />);
    wrapper.find(Button).at(1).simulate("click");
    expect(threadService.getLiked).toHaveBeenCalledTimes(1);
  });

  test("gets created threads", () => {
    const wrapper = shallow(<PersonalCabinetPage {...props} />);
    wrapper.find(Button).at(2).simulate("click");
    expect(threadService.getCreated).toHaveBeenCalledTimes(1);
  });

  test("gets replied threads", () => {
    const wrapper = shallow(<PersonalCabinetPage {...props} />);
    wrapper.find(Button).at(3).simulate("click");
    expect(threadService.getReplied).toHaveBeenCalledTimes(1);
  });
});
