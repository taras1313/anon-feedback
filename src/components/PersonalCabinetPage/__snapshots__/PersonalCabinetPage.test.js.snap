// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Personal cabinet page snapshot 1`] = `
<Fragment>
  <div
    className="buttonContainer"
  >
    <WithStyles(ForwardRef(ButtonGroup))
      aria-label="small outlined button group"
      color="primary"
      size="small"
    >
      <WithStyles(ForwardRef(Button))
        className="active"
        key="SUBSCRIBED"
        name="SUBSCRIBED"
        onClick={[Function]}
      >
        SUBSCRIBED
      </WithStyles(ForwardRef(Button))>
      <WithStyles(ForwardRef(Button))
        className=""
        key="LIKED"
        name="LIKED"
        onClick={[Function]}
      >
        LIKED
      </WithStyles(ForwardRef(Button))>
      <WithStyles(ForwardRef(Button))
        className=""
        key="CREATED"
        name="CREATED"
        onClick={[Function]}
      >
        CREATED
      </WithStyles(ForwardRef(Button))>
      <WithStyles(ForwardRef(Button))
        className=""
        key="REPLIED"
        name="REPLIED"
        onClick={[Function]}
      >
        REPLIED
      </WithStyles(ForwardRef(Button))>
    </WithStyles(ForwardRef(ButtonGroup))>
  </div>
  <Loader />
</Fragment>
`;
